extensions/amp-image-slider/0.1/test/integration/test-amp-image-slider.js

```diff
@@ -60,50 +66,760 @@ config.run('amp-image-slider', function() {
+              /*eventConfig*/
+              {
+                target: s1.slider,
+                cstr: createTouchMoveEvent,
+                pos: s1.pos.percent50,
+              },
+              /*observeTarget*/s1.slider,
+              /*sliderInfo*/s1,
+              /*targetPos*/s1.pos.percent50
+          );
+        });
+      });
+
+      it('should not move slider bar after touch lifted', () => {
+        // Touch start first to enable touchmove follow
+        s1.slider.dispatchEvent(createTouchStartEvent(s1.pos.percent50));
+        // Test to guarantee we are in touchmove mode
+        return verifyPositionUpdateAfterEventDispatch(
+            /*eventConfig*/
+            {
+              target: s1.slider,
+              cstr: createTouchMoveEvent,
+              pos: s1.pos.percent40,
+            },
+            /*observeTarget*/s1.slider,
+            /*sliderInfo*/s1,
+            /*targetPos*/s1.pos.percent40
+        ).then(() => {
+          // Release touch
+          s1.slider.dispatchEvent(createTouchEndEvent(s1.pos.percent40));
+          // Should not move now
+          s1.slider.dispatchEvent(createMouseMoveEvent(s1.pos.percent50));
+          return verifyPositionAfterTimeout(
+              /*sliderInfo*/s1, /*targetPos*/s1.pos.percent40);
+        });
+      });
+
+      it('slider bar should follow touchstart ' +
+        'after finishing previous touch', () => {
+        // Touch start first to enable touchmove follow
+        s1.slider.dispatchEvent(createTouchStartEvent(s1.pos.percent50));
+        // Test to guarantee we are in touchmove mode
+        return verifyPositionUpdateAfterEventDispatch(
+            /*eventConfig*/
+            {
+              target: s1.slider,
+              cstr: createTouchMoveEvent,
+              pos: s1.pos.percent40,
+            },
+            /*observeTarget*/s1.slider,
+            /*sliderInfo*/s1,
+            /*targetPos*/s1.pos.percent40
+        ).then(() => {
+          // Release touch
+          s1.slider.dispatchEvent(createTouchEndEvent(s1.pos.percent40));
+          // Should be able to trigger touchend again
+          return verifyPositionUpdateAfterEventDispatch(
+              /*eventConfig*/
+              {
+                target: s1.slider,
+                cstr: createTouchStartEvent,
+                pos: s1.pos.percent60,
+              },
+              /*observeTarget*/s1.slider,
+              /*sliderInfo*/s1,
+              /*targetPos*/s1.pos.percent60
+          );
+        });
+      });
+    });
+
+    describe('using a keyboard', () => {
+      it('pressing ArrowLeft should move slider bar by 10% to the left ' +
+        'only when slider is focused', () => {
+        // Notice that we are not focusing on the slider here
+        s1.slider.dispatchEvent(createKeyDownEvent('ArrowLeft'));
+        // Slider bar stays at original place (center by default)
+        return verifyPositionAfterTimeout(s1, s1.pos.percent50).then(() => {
```

Comment:
I think we only need one focus test to verify that without focus pressing a key doesn't do anything. That way, we can isolate each of these arrow tests to say, given focus, pressing key X moves slider bar to position Y, and the rest of the tests will not need chained promises.

Reply:
I am currently grouping focused and not focused tests for each key. Except for `Home` key that has special initial settings, each test should be just 2 promises chaining together now.